stan-dev/math
# Feature/issue 796 mpi cluster

#### Submission Checklist


#### Summary:
Add basic MPI cluster resource management.

#### Intended Effect:
Pave the way for map_rect_mpi.

#### How to Verify:
Ensure to have mpich or openMPI installed. Also add to `make/local`

```
CC=mpicxx
STAN_MPI=true
```

Then execute `./runTests.py test/unit/math/prim/arr/functor/mpi_cluster_test.cpp`. To run with more than just two processes please do `mpirun -np 4 test/unit/math/prim/arr/functor/mpi_cluster_test`.

#### Side Effects:

#### Documentation:
Doxygen documentation in the code.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
